File: src/Asm.MooBank.Modules.Bills/Models/Bill.cs
namespace Asm.MooBank.Modules.Bills.Models;

public record Bill : BillBase
{
    public int Id { get; init; }

    public Guid AccountId { get; init; }

    public string AccountName { get; init; } = String.Empty;
}

public static class BillExtensions
{
    public static Bill ToModel(this Asm.MooBank.Domain.Entities.Utility.Bill entity)
    {
        return new Bill
        {
            Id = entity.Id,
            AccountId = entity.AccountId,
            AccountName = entity.Account?.Name ?? String.Empty,
            InvoiceNumber = entity.InvoiceNumber,
            IssueDate = entity.IssueDate,
            CurrentReading = entity.CurrentReading,
            PreviousReading = entity.PreviousReading,
            Total = entity.Total,
            CostsIncludeGST = entity.CostsIncludeGST,
            Cost = entity.Cost,
            Periods = entity.Periods.ToModel(),
            Discounts = entity.Discounts.ToModel(),
        };
    }

    public static IEnumerable<Bill> ToModel(this IEnumerable<Asm.MooBank.Domain.Entities.Utility.Bill> entities)
    {
        return entities.Select(entity => entity.ToModel());
    }
}

File: src/Asm.MooBank.Web.App/src/models/bills/Bill.ts
import { Period } from "./Period";

export interface BillBase {
    invoiceNumber: string;
    issueDate: string;
    currentReading?: number;
    previousReading?: number;
    total: number;
    costsIncludeGST?: boolean;
    cost: number;
    periods?: Period[];
}

export interface Bill extends BillBase {
    id: number;
    accountId: string;
    accountName: string;
}

export interface CreatePeriod {
    periodStart: string;
    periodEnd: string;
    pricePerUnit: number;
    totalUsage: number;
    chargePerDay: number;
}

export interface CreateDiscount {
    discountPercent?: number;
    discountAmount?: number;
    reason?: string;
}

export interface CreateBill {
    invoiceNumber?: string;
    issueDate: string;
    currentReading?: number;
    previousReading?: number;
    total: number;
    costsIncludeGST?: boolean;
    cost: number;
    periods: CreatePeriod[];
    discounts: CreateDiscount[];
}